### ISSUE TITLE
TypeChecking/Substitute.hs:107  in Agda-oct19-2015

### ISSUE THREAD

**Issue by mechvel** [agda/agda#1695]
Development Agda of October 19, 2015 
applied as
             `agda $agdaLibOpt +RTS ... -RTS  --sharing  ...`

reports

...
`Skipping List.L1 (/home/mechvel/doconA/0.04/source/List/L1.agdai).`

`conApp: constructor Relation.Nullary.Dec.yes with fields [] projected by` 
`Relation.Binary.DecSetoid.Carrier`

`An internal error has occurred. Please report this as a bug.`
`Location of the error: src/full/Agda/TypeChecking/Substitute.hs:107`

This looks similar to certain recent bug  of  `--sharing`  related to a constant operation,
which has been fixed recently.
But the example is large.
May I provide it as it is, without reducing it?

**Comment by andreasabel** [comment]
For now, yes, please provide the large example.  (We might ask for your help in shrinking later.)

**Comment by mechvel** [comment]
It occurs easy to reduce to a small thing.
Please, download it from

   http://www.botik.ru/pub/local/Mechveliani/agdaNotes/Bug-oct20-2015.agda

It reports 

`conApp: constructor Relation.Nullary.Dec.yes with fields [] projected by` `Relation.Binary.DecSetoid.Carrier`
`An internal error has occurred. Please report this as a bug.`
`Location of the error: src/full/Agda/TypeChecking/Substitute.hs:107`

both for  --no-sharing  and  --sharing.

**Comment by asr** [comment]
Duplicate of https://github.com/agda/agda/issues/1407, so I'm closing this one.

**Comment by andreasabel** [comment]
@mechvel, I'd like to test your file linked above. But the link no longer works, and http://web.archive.org does not have it due to your http://www.botik.ru/robots.txt
Can you (preferably) post the contents of this file here?  (Or reactivate the link.)

**Comment by mechvel** [comment]
All right, I return this file there.
Please, try to dowload it once more.

**Comment by mechvel** [comment]
``` agda
-- It fails both with  --sharing  and  --no-sharing
-- It is similar to  Isuue 1407.

module Bug-oct20-2015 where

open import Level using (Level; _⊔_) 
open import Function using (case_of_)
open import Relation.Nullary using (¬_; Dec; yes; no)
open import Relation.Binary  using (DecSetoid; module DecSetoid)
open import Relation.Binary.PropositionalEquality as PE using (_≡_)
open import Data.Empty using (⊥; ⊥-elim)
open import Data.Product     using (_×_; proj₁; proj₂; _,_) 
open import Data.List using (List; []; _∷_; map) 
open import Data.List.Any using (Any; module Any; module Membership) 

open Any 

module OfAssocList {α α= β : Level} (Key : DecSetoid α α=) (Val : Set β) 
  where
  open DecSetoid Key using (_≈_; _≟_)
                     renaming (Carrier to K; setoid to setoidK)
  KV = K × Val

  open Membership setoidK using (_∈_) 

  Pairs : Set _
  Pairs = List KV

  getKeys : Pairs → List K
  getKeys = map proj₁ 

  CombineKVV : Set (α ⊔ β)
  CombineKVV = K → Val → Val → Val

  insertWithKey : CombineKVV →  KV → Pairs → Pairs
  insertWithKey _    p         []                =  p ∷ []
  insertWithKey comb (k , new) ((k' , old) ∷ ps) =
        case k ≟ k' 
        of \ 
        { (yes _) → (k' , comb k new old) ∷ ps
        ; (no _)  → (k' , old) ∷ (insertWithKey comb (k , new) ps) 
        }


  keys-insertWithKey-eq-if∈ : (comb : CombineKVV) → (p : KV) → (ps : Pairs) → 
                               let ins-ps = insertWithKey comb p ps
                                   keys   = getKeys ps
                                   k      = proj₁ p
                               in
                               k ∈ keys → getKeys ins-ps ≡ keys

  keys-insertWithKey-eq-if∈ _ _       []              () 
  keys-insertWithKey-eq-if∈ _ (k , _) ((k' , _) ∷ ps) (here k=k')  with k ≟ k'
  ...          
      | yes _   =  PE.refl
  ... | no k≉k' =  ⊥-elim (k≉k' k=k') 

  keys-insertWithKey-eq-if∈ comb (k , v) ((k' , v') ∷ ps) (there k∈ks)  
                                                                 with k ≟ k' 
  ... | yes k=k' =  PE.refl
  ... | no k≉k'  =  PE.cong (_∷_ k')  
                            (keys-insertWithKey-eq-if∈ comb (k , v) ps k∈ks)
```

**Comment by andreasabel** [comment]
This works after the fix of #1407, so I guess it was indeed a duplicate.
